### PR TITLE
imageRepoDigest

### PR DESCRIPTION
**Pull request by cyastella**
### Summary

To fetch the image RepoDigest when inspect the image to help find where the image pull from.
run docker images --digests 
### Testing

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/engine/docker_image_manager.go [hunk 1]
```diff
@@ -126,15 +128,34 @@ func (imageManager *dockerImageManager) RecordContainerReference(container *apic
...
+	imageRepoDigests := imageInspected.RepoDigests
+	resultRepoDigest := ""
+	imagePrefix := strings.Split(container.Image, ":")[0]
+	for _, imageRepoDigest := range imageRepoDigests {
+		if strings.HasPrefix(imageRepoDigest, imagePrefix) {
+			repoDigestSplitList := strings.Split(imageRepoDigest, "@")
+			if len(repoDigestSplitList) > 1 {
+				resultRepoDigest = repoDigestSplitList[1]
+			} else {
+				seelog.Debugf("imageRepoDigest doesn't have the right format: [%s]", imageRepoDigest)
```
Comment (suneyz) [inline agent/engine/docker_image_manager.go]: You might want to log this as a warn when it doesn't have the format you are expecting.

File: agent/engine/docker_image_manager.go [hunk 2]
```diff
@@ -126,15 +128,34 @@ func (imageManager *dockerImageManager) RecordContainerReference(container *apic
...
+// The helper function to fetch the RepoImageDigest when inspect the image
+func (imageManager *dockerImageManager) fetchRepoDigest(imageInspected *types.ImageInspect, container *apicontainer.Container) string {
+	imageRepoDigests := imageInspected.RepoDigests
+	resultRepoDigest := ""
+	imagePrefix := strings.Split(container.Image, ":")[0]
+	for _, imageRepoDigest := range imageRepoDigests {
+		if strings.HasPrefix(imageRepoDigest, imagePrefix) {
+			repoDigestSplitList := strings.Split(imageRepoDigest, "@")
+			if len(repoDigestSplitList) > 1 {
+				resultRepoDigest = repoDigestSplitList[1]
```
Comment (sharanyad) [inline agent/engine/docker_image_manager.go]: I think you can just break here to return the digest.

File: agent/engine/docker_image_manager_test.go [hunk 3]
```diff
@@ -272,6 +272,74 @@ func TestAddContainerReferenceToExistingImageState(t *testing.T) {
...
+			},
+			imageInspected: &types.ImageInspect{
+				RepoDigests: []string{"mysql", "repo2@digest2"},
+			},
+		},
+	}
+
+	for i, test := range testCases {
+		resultRepoDigest := imageManager.fetchRepoDigest(test.imageInspected, test.container)
+		assert.Equal(t, test.container.ImageDigest, resultRepoDigest, "Error has the right repoDigest", i)
```
Comment (sharanyad) [inline agent/engine/docker_image_manager_test.go]: nit: you can change this error message to `incorrect repoDigest` or similar.

File: agent/engine/docker_image_manager.go (hunk 1 above)
Comment (sharanyad) [inline agent/engine/docker_image_manager.go]: nit: `ImageRepoDigest`

File: agent/api/statechange.go [hunk 4]
```diff
@@ -144,10 +146,10 @@ func NewContainerStateChangeEvent(task *apitask.Task, cont *apicontainer.Contain
 		Status:        contKnownStatus.BackendStatus(cont.GetSteadyStateStatus()),
 		ExitCode:      cont.GetKnownExitCode(),
 		PortBindings:  cont.GetKnownPortBindings(),
+		ImageDigest:   cont.ImageDigest,
```
Comment (sharanyad) [inline agent/api/statechange.go]: let's use `GetImageDigest()` with a lock, just in case.

File: agent/api/container/container.go [hunk 5]
```diff
@@ -110,6 +110,8 @@ type Container struct {
 	Image string
 	// ImageID is the local ID of the image used in the container
 	ImageID string
+	// ImageDigest is the sha-256 digest of the container image as pulled from the repository
+	ImageDigest string
```
Comment (sharanyad) [inline agent/api/container/container.go]: this may require a statefile version update

File: agent/api/statechange.go (hunk 4 above)
Comment (cyastella) [reply agent/api/statechange.go]: cool, will change it.

File: agent/api/container/container.go (hunk 5 above)
Comment (cyastella) [reply agent/api/container/container.go]: will update the statefile version.

**Pull request merged**